keyvault/client/src/main/java/com/azure/keyvault/SecretClient.java

```diff
@@ -0,0 +1,540 @@
+
+        return getPagedSecrets(result);
+    }
+
+
+    /**
+     * Lists {@link DeletedSecret} secrets of the key vault.
+     * The Get Deleted Secrets operation returns the secrets that have been deleted for a vault enabled for soft-delete.
+     * This operation requires the secrets/list permission.
+     *
+     * @param maxPageResults Maximum number of results to return in a page. If not specified, the service will return up to 25 results.
+     * @throws NullPointerException thrown if maxPageResults parameter is null
+     * @return A Flux of Deleted Secrets. The Flux contains all of the deleted secrets in the vault.
+     */
+    public Flux<DeletedSecret> listDeletedSecretsAsync(int maxPageResults) {
+        Objects.requireNonNull(maxPageResults, "The maximum page results parameter cannot be null.");
+        Mono<RestResponse<Page<DeletedSecret>>> result = service.getDeletedSecrets(vaultEndPoint, maxPageResults, API_VERSION, ACCEPT_LANGUAGE);
+        return getPagedDeletedSecrets(result);
+    }
+
+    /**
+     * List all versions of the specified secret.
+     * The full secret identifier and attributes are provided in the response. No values are returned for the secrets.
+     * This operations requires the secrets/list permission.
+     *
+     * @param name The name of the secret.
+     * @param maxPageResults Maximum number of results to return in a page. If not specified, the service will return up to 25 results.
+     * @throws NullPointerException thrown if {@code name} or {@code maxPageResults} is null
+     * @return A Flux of Secrets. The Flux contains all of the versions of the secret in the vault.
+     */
+    public Flux<SecretInfo> listSecretVersionsAsync(String name, int maxPageResults) {
+        Objects.requireNonNull(name, "The Secret name cannot be null.");
+        Objects.requireNonNull(maxPageResults, "The maximum page results parameter cannot be null.");
+        Mono<RestResponse<Page<SecretInfo>>> result = service.getSecretVersions(vaultEndPoint, name, maxPageResults, API_VERSION, ACCEPT_LANGUAGE);
+
+        return getPagedSecrets(result);
+    }
+
+    private Flux<SecretInfo> listSecretsNext(@NonNull String nextPageLink) {
+        Mono<RestResponse<Page<SecretInfo>>> result = service.getSecrets(vaultEndPoint, nextPageLink, ACCEPT_LANGUAGE);
+        return getPagedSecrets(result);
+    }
+
+    private Flux<DeletedSecret> listDeletedSecretsNext(@NonNull String nextPageLink) {
+        Mono<RestResponse<Page<DeletedSecret>>> result = service.getDeletedSecrets(vaultEndPoint, nextPageLink, ACCEPT_LANGUAGE);
+        return getPagedDeletedSecrets(result);
+    }
+
+    private Flux<SecretInfo> getPagedSecrets(Mono<RestResponse<Page<SecretInfo>>> response) {
+        return response.flatMapMany(p -> Flux.just(new RestPagedResponseImpl<>(p.body().items(), p.body().nextPageLink(), p.request(), p.headers(), p.statusCode())))
+                .concatMap(this::extractAndFetchSecrets);
+    }
```

Comment:
Super keen to see this extracted out into azure-common implementation API.

Reply:
Noted, needs to be moved to Azure common.